Classes/Plugin/Tools/SearchInDocumentTool.php

```diff
@@ -113,14 +105,68 @@ public function main($content, $conf)
             '###LABEL_PREVIOUS###' => htmlspecialchars($this->pi_getLL('label.previous')),
             '###LABEL_PAGE###' => htmlspecialchars($this->pi_getLL('label.logicalPage')),
             '###LABEL_NORESULT###' => htmlspecialchars($this->pi_getLL('label.noresult')),
-            '###CURRENT_DOCUMENT###' => $this->doc->uid,
-            '###SOLR_ENCRYPTED###' => $encryptedSolr ?: '',
+            '###LABEL_QUERY_URL###' => $this->conf['queryInputName'],
+            '###LABEL_START###' => $this->conf['startInputName'],
+            '###LABEL_ID###' => $this->conf['idInputName'],
+            '###LABEL_PAGE_URL###' => $this->conf['pageInputName'],
+            '###LABEL_HIGHLIGHT_WORD###' => $this->conf['highlightWordInputName'],
+            '###LABEL_ENCRYPTED###' => $this->conf['encryptedInputName'],
+            '###CURRENT_DOCUMENT###' => $this->getCurrentDocumentId(),
+            '###SOLR_ENCRYPTED###' => $this->getEncryptedCoreName() ?: '',
         ];
 
         $content .= $this->templateService->substituteMarkerArray($this->template, $markerArray);
         return $this->pi_wrapInBaseClass($content);
     }
 
+    /**
+     * Get the action url for search form
+     *
+     * @access protected
+     *
+     * @return string with action url for search form
+     */
+    protected function getActionUrl()
+    {
+        // Configure @action URL for form.
+        $linkConf = [
+            'parameter' => $GLOBALS['TSFE']->id,
+            'forceAbsoluteUrl' => 1,
+            'forceAbsoluteUrl.' => ['scheme' => !empty($this->conf['forceAbsoluteUrlHttps']) ? 'https' : 'http']
+        ];
+
+        $actionUrl = $this->cObj->typoLink_URL($linkConf);
+
+        if (!empty($this->conf['searchUrl'])) {
+            $actionUrl = $this->conf['searchUrl'];
+        }
+        return $actionUrl;
+    }
+
+    /**
+     * Get current document id
+     *
+     * @access protected
+     *
+     * @return string with current document id
+     */
+    protected function getCurrentDocumentId()
+    {
+        $id = $this->doc->uid;
+
+        if (!empty($this->conf['documentIdUrlSchema'])) {
```

Comment:
What's the purpose of this? Could you please provide an example string to verify the string splitting magic below?

Reply:
Sometimes id of the document is actually not an id but a link to xml file. Then to search in the index the full link can't be used so it is needed to extract real id out of the full link.

Example:
https://dev-ddb.fiz-karlsruhe.de/api/items/F45KF24U7UIIVCUJJBV334PD7IPE7JBW/source/record -> https://dev-ddb.fiz-karlsruhe.de/api/items/*id*/source/record

Reply:
I think we need to distinguish more cases here:
- The normal case would be documents which are properly indexed in the TYPO3 database. Those don't need any special treatment for their ID anyway.
- The "DFG-Viewer" case would be documents which are not indexed in the database **nor in the Solr index**. Normally those can be identified by the fact that they have an URL as ID.
- Then there is the special "Zeitungsportal" case where we have documents, which are not indexed in the database, but are indexed in the Solr index. Those have an URL as ID as well, but need special treatment to extract the indexed ID from the URL.

I suggest using the TYPO3 hook mechanism to handle all those cases:
- If a document has a numerical ID, no further treatment is necessary.
- If a document has a string ID, this string is processed through a hook.
  - If the hook returns an integer, the document is treated as indexed.
  - If the hook returns the string unchanged, the document is considered unindexed and ignored when searching. (This would be the default if no hook is registered, e. g. in "DFG-Viewer mode".)

The advantages would be:
- We don't need to add Zeitungportal-specific logic to the main extension. Instead the hook can be part of the supplemental Zeitungsportal extension.
- The solution is extensible, because you can register more than one hook to handle different kinds of identifiers. So anybody with a setup of a pre-existing Solr index could map their document IDs to the index IDs and re-use their Solr index. (This could even be interesting for SLUB, when we think about consolidating our different indices - CMR, Find, Kitodo, ...)
- Processing the ID string wouldn't be limited to splitting a string, because a hook could include more complex logic like regular expressions or even database lookups.

Reply:
> I suggest using the TYPO3 hook mechanism to handle all those cases:

Do you mean https://typo3.org/article/how-to-use-existing-hooks-in-your-own-extension ?

> If a document has a string ID, this string is processed through a hook.
> If the hook returns an integer, the document is treated as indexed.
> If the hook returns the string unchanged, the document is considered unindexed and ignored when searching. (This would be the default if no hook is registered, e. g. in "DFG-Viewer mode".)

But how to convert string ID to an integer?

Reply:
We actually already have some hooks in Kitodo.Presentation. A good example is the way how we extract the record ID from METS:
Normally we would take `/mets:mets[@OBJID]` as the document's unique record identifier: https://github.com/kitodo/kitodo-presentation/blob/f12eede1596112dc359678768e9a9463f0791788/Classes/Common/MetsDocument.php#L147-L150
But if this attribute is not set (which is the case for old versions of Kitodo.Production), we use hooks to have a fallback. The hook is registered in `ext_localconf.php`: https://github.com/kitodo/kitodo-presentation/blob/f12eede1596112dc359678768e9a9463f0791788/ext_localconf.php#L219 and called in `Classes/Common/MetsDocument.php`: https://github.com/kitodo/kitodo-presentation/blob/f12eede1596112dc359678768e9a9463f0791788/Classes/Common/MetsDocument.php#L151-L158
The hook itself can be found in `Classes/Hooks/KitodoProductionHacks.php`: https://github.com/kitodo/kitodo-presentation/blob/f12eede1596112dc359678768e9a9463f0791788/Classes/Hooks/KitodoProductionHacks.php#L37-L68

Reply:
Hook is a possibility. I think the modern way would be to make some "middleware". But I haven't done this before. Hook support itself seem to be not deprecated, although lots of core hooks are.

The DFG-Viewer case is not relevant in the SearchInDocument tool. This tool won't work with it anyway. The DFG-Viewer has it's own SRU search plugin, which requires a SRU server of course.

I suggest to first test and documentate this PR including indexing (doesn't work yet) and searching for Kitodo.Presentation alone. If this works, we can discuss further at this point.